### PR TITLE
Working on shell scopes.

### PR DESCRIPTION
**Pull request by jtkech**
Here the goals are

- To be able to retrieve the current shell scope (if it exists) from the async flow.

- The above allowed me to remove some usages of `HttpContext.RequestServices`.

- Shell scope `UsingAsync()` helper which also manages e.g shell activation, deferred tasks.

- Allow to register `onCompleted` callbacks that are executed through the above helper.

- Here, i used an `onCompleted` callback to commit the `ISession`.

### CODE REVIEW

**Comment by jtkech** [comment]
Just for infos

- So, we have a `ShellScope` holding a `ShellContext` and a scope `IServiceProvider`, and that we can retrieve during all the local async execution flow of a given shell scope. This without having to swap any state when creating / disposing a shell scope.

- This allowed me to remove some usages of `RequestServices`. Maybe i could remove more usages but it makes sense e.g in some of our middlewares and it is used in some mvc sevices ... That's why i first kept the swapping of `HttpContext.RequestServices` when creating / disposing a shell scope.

- But i created a `ShellScopeServices` wrapper which uses the services of the current shell scope if it exists. This wrapper allows to make `RequestServices` aware of the current `ShellScope`, so that we don't need to swap it anymore. **Normally this allows to have interleaved shell scopes**.

- The `ShellScope` can also manages shell activations and deferred tasks. We can also register calbacks to be invoked just before disposing a shell scope, e.g to commit the session.

- To make all the above working, the only constraint is that in place of using e.g this.

      var shellScope = await _shellHost.GetScopeAsync(shellSettings);

      using (shellScope)
      {
          ....
      }

    The constraint is to use this pattern.

      var shellScope = await _shellHost.GetScopeAsync(shellSettings);

      await shellScope.UsingAsync(scope =>
      {
          ....
      });

**Comment by Skrypt** [comment]
Have you tried running concurrency calls to a YesSQL session with this PR to see if it would not fix the issues that we're reported by @jpiquot ?

**Comment by jtkech** [comment]
Hmm, i hope but not sure.

As i remember, what i tried is to create concurrently shell scopes which then use yesSql but each one with their own session. And it was working because here we don't swap the `RequestServices` explicitly, we use another way retrieving an `AsyncLocal` scope. Also we commit the session before disposing the scope.

**Comment by jtkech** [comment]
Maybe worth to be reviewed.

- E.g we can commit the session before releasing the scope.

- E.g much more easier to retrieve a shell scope and its service provider. Note: many usage of httpContext.RequestServices may be replaced (but not necessarily).

It will help me when working again on distributed services.

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        /// Activate the shell, if not yet done, by calling the related tenant event handlers.
+        /// </summary>
+        public async Task ActivateShellAsync()
+        {
+            if (ShellContext.IsActivated)
+            {
+                return;
+            }
+
+            var semaphore = _semaphores.GetOrAdd(ShellContext.Settings.Name, (name) => new SemaphoreSlim(1));
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: This needs to be a `lock` around a standard dictionary, and also a lock in the `Remove` below.

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+using OrchardCore.Modules;
+
+namespace OrchardCore.Environment.Shell.Scope
+{
+    /// <summary>
+    /// Custom 'IServiceScope' managing the shell state and the execution flow.
+    /// </summary>
+    public class ShellScope : IServiceScope
+    {
+        private static AsyncLocal<ShellScope> _current = new AsyncLocal<ShellScope>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: `readonly`

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+        private static readonly ConcurrentDictionary<string, SemaphoreSlim> _semaphores =
+            new ConcurrentDictionary<string, SemaphoreSlim>();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: `Dictionary` (see below)

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+                ShellContext.Dispose();
+            }
+
+            // Decrement the counter at the very end of the scope
+            Interlocked.Decrement(ref ShellContext._refCount);
+
+            _disposed = true;
+        }
+
+        public void Dispose() => DisposeAsync().GetAwaiter().GetResult();
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: Checked `_disposed` before calling `GetAwaiter().GetResult()` so it's not used in the general case

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs (hunk 1 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: Okay.

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs (hunk 2 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: Okay.

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs (hunk 3 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: Okay.

File: src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs (hunk 4 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs]: For sure.

**Comment by jtkech** [comment]
@sebastienros 

I replaced the concurrent dictionary to a dictionary + lock. I was using a concurrent dico as it was done in the `ModularTenantContainerMiddleware.cs`.

If we start many tenants concurrently, we will have many reads and updates, are you sure that in this case it is better to use a dictionary + lock?

If it is better why not doing the same in `ModularTenantRouterMiddleware.cs`?

**Pull request merged**